File: src/exoticatechnologies/modifications/exotics/impl/TierIIIDriveSystem.java
package exoticatechnologies.modifications.exotics.impl;

import com.fs.starfarer.api.campaign.CampaignFleetAPI;
import com.fs.starfarer.api.campaign.econ.MarketAPI;
import com.fs.starfarer.api.combat.MutableShipStatsAPI;
import com.fs.starfarer.api.fleet.FleetMemberAPI;
import com.fs.starfarer.api.fleet.MutableFleetStatsAPI;
import com.fs.starfarer.api.ui.TooltipMakerAPI;
import com.fs.starfarer.api.ui.UIComponentAPI;
import exoticatechnologies.modifications.ShipModifications;
import exoticatechnologies.modifications.exotics.Exotic;
import exoticatechnologies.modifications.exotics.ExoticData;
import exoticatechnologies.util.StringUtils;
import exoticatechnologies.util.Utilities;
import lombok.Getter;
import org.json.JSONObject;

import java.awt.*;
import java.util.HashMap;
import java.util.Map;

public class TierIIIDriveSystem extends Exotic {
    private static float CARGO_TO_FUEL_PERCENT = 75f;
    private static float BURN_BONUS_FUEL_REQ = 66f;
    private static float BURN_BONUS = 2f;

    @Getter
    private final Color color = new Color(0xFFA2E4);

    public TierIIIDriveSystem(String key, JSONObject settings) {
        super(key, settings);
    }

    @Override
    public boolean canAfford(CampaignFleetAPI fleet, MarketAPI market) {
        return Utilities.hasItem(fleet.getCargo(), ITEM);
    }

    @Override
    public boolean removeItemsFromFleet(CampaignFleetAPI fleet, FleetMemberAPI fm, MarketAPI market) {
        Utilities.takeItemQuantity(fleet.getCargo(), ITEM, 1);

        return true;
    }

    @Override
    public Map<String, Float> getResourceCostMap(FleetMemberAPI fm, ShipModifications mods, MarketAPI market) {
        Map<String, Float> resourceCosts = new HashMap<>();
        resourceCosts.put(
                "&" + StringUtils.getTranslation("ShipListDialog", "ChipName")
                        .format("name", getName())
                        .toStringNoFormats(), 1f);
        return resourceCosts;
    }

    @Override
    public void modifyToolTip(TooltipMakerAPI tooltip, UIComponentAPI title, FleetMemberAPI fm, ShipModifications mods, boolean expand) {
        if (expand) {
            StringUtils.getTranslation(this.getKey(), "longDescription")
                    .format("cargoToFuelPercent", CARGO_TO_FUEL_PERCENT)
                    .format("burnBonusFuelReq", BURN_BONUS_FUEL_REQ)
                    .format("burnBonus", BURN_BONUS)
                    .addToTooltip(tooltip, title);
        }
    }

    @Override
    public void applyExoticToStats(String id, FleetMemberAPI fm, MutableShipStatsAPI stats, ExoticData data) {
        int addedFuelCap = (int) (fm.getCargoCapacity() * CARGO_TO_FUEL_PERCENT / 100f);
        stats.getCargoMod().modifyMult(this.getBuffId(), 1 - (CARGO_TO_FUEL_PERCENT / 100f));
        stats.getFuelMod().modifyFlat(this.getBuffId(), addedFuelCap);
    }

    @Override
    public void advanceInCampaign(FleetMemberAPI member, ShipModifications mods, Float amount, ExoticData data) {
        if (member.getFleetData() != null && member.getFleetData().getFleet() != null) {
            checkBuff(member.getFleetData().getFleet());

        }
    }

    @Override
    public void onDestroy(FleetMemberAPI member) {
        if (member.getFleetData() != null && member.getFleetData().getFleet() != null) {
            removeBuff(member.getFleetData().getFleet().getStats());
        }
    }

    private void checkBuff(CampaignFleetAPI fleet) {
        MutableFleetStatsAPI fleetStats = fleet.getStats();
        if (fleet.getCargo().getFreeFuelSpace() < fleet.getCargo().getMaxFuel() * 0.25) {
            if (fleetStats.getFleetwideMaxBurnMod().getFlatBonus(this.getBuffId()) == null) {
                fleetStats.getFleetwideMaxBurnMod().modifyFlat(this.getBuffId(), BURN_BONUS, this.getName());
            }
        } else {
            removeBuff(fleetStats);
        }
    }

    private void removeBuff(MutableFleetStatsAPI fleetStats) {
        if (fleetStats.getFleetwideMaxBurnMod().getFlatBonus(this.getBuffId()) != null) {
            fleetStats.getFleetwideMaxBurnMod().unmodify(this.getBuffId());
        }
    }
}

File: src/exoticatechnologies/modifications/exotics/impl/TierIIIDriveSystem.kt
package exoticatechnologies.modifications.exotics.impl

import com.fs.starfarer.api.campaign.CampaignFleetAPI
import com.fs.starfarer.api.campaign.econ.MarketAPI
import com.fs.starfarer.api.combat.MutableShipStatsAPI
import com.fs.starfarer.api.fleet.FleetMemberAPI
import com.fs.starfarer.api.impl.campaign.ids.Stats
import com.fs.starfarer.api.ui.TooltipMakerAPI
import com.fs.starfarer.api.ui.UIComponentAPI
import exoticatechnologies.modifications.ShipModifications
import exoticatechnologies.modifications.exotics.Exotic
import exoticatechnologies.modifications.exotics.ExoticData
import exoticatechnologies.util.StringUtils
import exoticatechnologies.util.Utilities
import org.json.JSONObject
import java.awt.Color

class TierIIIDriveSystem(key: String, settings: JSONObject) : Exotic(key, settings) {
    override var color = Color(0xFFA2E4)
    override fun canAfford(fleet: CampaignFleetAPI, market: MarketAPI?): Boolean {
        return Utilities.hasItem(fleet.cargo, ITEM)
    }

    override fun removeItemsFromFleet(fleet: CampaignFleetAPI, member: FleetMemberAPI, market: MarketAPI?): Boolean {
        Utilities.takeItemQuantity(fleet.cargo, ITEM, 1f)
        return true
    }

    private fun getCargoToFuelPercent(member: FleetMemberAPI, mods: ShipModifications, exoticData: ExoticData): Float {
        return CARGO_TO_FUEL_PERCENT + (100 - CARGO_TO_FUEL_PERCENT) * (getNegativeMult(member, mods, exoticData) - 1f)
    }

    override fun modifyToolTip(
        tooltip: TooltipMakerAPI,
        title: UIComponentAPI,
        member: FleetMemberAPI,
        mods: ShipModifications,
        exoticData: ExoticData,
        expand: Boolean
    ) {
        if (expand) {
            StringUtils.getTranslation(key, "longDescription")
                .format("cargoToFuelPercent", getCargoToFuelPercent(member, mods, exoticData))
                .format("burnBonusFuelReq", BURN_BONUS_FUEL_REQ / getPositiveMult(member, mods, exoticData))
                .format("burnBonus", BURN_BONUS)
                .addToTooltip(tooltip, title)
        }
    }

    override fun applyExoticToStats(
        id: String,
        stats: MutableShipStatsAPI,
        member: FleetMemberAPI,
        mods: ShipModifications,
        exoticData: ExoticData
    ) {
        val addedFuelCap = (member.cargoCapacity * getCargoToFuelPercent(member, mods, exoticData) / 100f).toInt()
        stats.cargoMod.modifyMult(buffId, 1 - getCargoToFuelPercent(member, mods, exoticData) / 100f)
        stats.fuelMod.modifyFlat(buffId, addedFuelCap.toFloat())

        stats.dynamic.getMod(Stats.FLEET_BURN_BONUS).modifyFlat(buffId, BURN_BONUS, name)
    }

    companion object {
        private const val CARGO_TO_FUEL_PERCENT = 75f
        private const val BURN_BONUS_FUEL_REQ = 66f
        private const val BURN_BONUS = 2f
    }
}